File: 1626.无矛盾的最佳球队.go
/*
 * @lc app=leetcode.cn id=1626 lang=golang
 *
 * [1626] 无矛盾的最佳球队
 */

// @lc code=start
func bestTeamScore(scores []int, ages []int) (res int) {
	length := len(ages)
	ageSortedScores := make([][2]int, length)
	for idx, val := range ages {
		ageSortedScores[idx] = [2]int{val, scores[idx]}
	}
	ageSortedScores = append([][2]int{{0, 0}}, ageSortedScores...) // 添加一个哨兵
	sort.Slice(ageSortedScores, func(i, j int) bool {
		if ageSortedScores[i][0] == ageSortedScores[j][0] {
			return ageSortedScores[i][1] < ageSortedScores[j][1]
		}
		return ageSortedScores[i][0] < ageSortedScores[j][0]
	})
	dp := make([]int, length+1)
	for i := 1; i < length+1; i++ {
		for j := 0; j < i; j++ {
			// 遍历所有同龄或者低龄者，需满足分数小于等于当前球员
			if ageSortedScores[i][1] >= ageSortedScores[j][1] {
				// 获取最长增长子序列
				dp[i] = max(dp[i], ageSortedScores[i][1]+dp[j])
				res = max(res, dp[i])
			}
		}
	}
	return
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// @lc code=end

File: 1626.无矛盾的最佳球队.py
#
# @lc app=leetcode.cn id=1626 lang=python3
#
# [1626] 无矛盾的最佳球队
#

# @lc code=start
class Solution:
    def bestTeamScore(self, scores: List[int], ages: List[int]) -> int:
        age_sorted_scores = [score for _, score in sorted(zip(ages, scores))]
        age_sorted_scores = [0] + age_sorted_scores # 哨兵，简化，持续和之前的所有年龄小于等于者以及分数小于者做序列总和比较
        dp = [0] * len(age_sorted_scores)
        for i in range(1, len(age_sorted_scores)):
            for j in range(i):
                if age_sorted_scores[i] >= age_sorted_scores[j]:
                    dp[i] = max(dp[i], dp[j] + age_sorted_scores[i]) # 取以 i 位置的最大和
        return max(dp)

# @lc code=end

